### PR TITLE
Momentum Diffusion

### PR DESCRIPTION
**Pull request by lukasmerten**
This is a rewrite of the content of PR #279, which will be closed.

The module introduces a momentum diffusion module, that allows to treat 2nd order Fermi acceleration in the ensemble averaged picture. It is meant to be used together with the DiffusionSDE module. In the future the module can be extended to include more complicated dependencies of the momentum diffusion scalar.

Unit tests and documentation / example code are included.

### CODE REVIEW

File: src/module/MomentumDiffusion.cpp [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	setDpp(Dpp);
+}
+
+void ConstantMomentumDiffusion::process(Candidate *c) const {
+	double rig = c->current.getRigidity();
+	if (std::isinf(rig)) {
+		return; // Only charged particles
+	}
+	
+	double p = c->current.getEnergy()/c_light; // Note we use E=p/c (relativistic limit)
```
Comment (rafaelab) [inline src/module/MomentumDiffusion.cpp:16]: add spaces:
```getEnergy() / c_light```

File: src/module/MomentumDiffusion.cpp [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	double dt = c->getCurrentStep() / c_light;
+	
+	double eta =  Random::instance().randNorm();
+	double domega = eta * sqrt(dt);
+	
+	double AScal = calculateAScalar(p);
+	double BScal = calculateBScalar();
+
+	double dp = AScal * dt + BScal * domega;
+	c->current.setEnergy((p + dp)*c_light);
```
Comment (rafaelab) [inline src/module/MomentumDiffusion.cpp:26]: Add spaces between parenthesis and c_light.

File: src/module/MomentumDiffusion.cpp [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	
+	c->limitNextStep(limit * p / AScal * c_light);
+}
+
+double ConstantMomentumDiffusion::calculateAScalar(double p) const {
+	double a = + 2./p * Dpp;
```
Comment (rafaelab) [inline src/module/MomentumDiffusion.cpp:32]: Follow CRPropa convention and add white spaces around  `/`

File: src/module/MomentumDiffusion.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+double ConstantMomentumDiffusion::getLimit() const {
+	return limit;
+}
+
+std::string ConstantMomentumDiffusion::getDescription() const {
+	std::stringstream s;
+	s << "limit: " << limit << "\n";
+	s << "Dpp: " << Dpp / (meter*meter/second) << " m^2/s";
```
Comment (rafaelab) [inline src/module/MomentumDiffusion.cpp:63]: Add white spaces, following CRPropa's convention.

**Comment by rafaelab** [comment]
Hi @lukasmerten 
Thanks for this update. It runs just fine in my laptop, and so does the example notebook.
I left a few (very minor) comments regarding CRPropa's style convention.
I also left a comment about the `limit` option in this module, if we want to leave it as a keyword argument with default value as in other modules.

**Comment by lukasmerten** [comment]
@rafaelab I adressed all your style convention comments and added a new constructor that allows for direct setting of the limit parameter. Let me know if there is anything else that needs to be done.

**Review by rafaelab — APPROVED** [review]

**Pull request merged**